### ISSUE TITLE
Pathological GC time on M1 mac

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#48473]
M1 mac:
```
category = "linked"
bench = "list.jl"
  No Changes to `~/GCBenchmarks/benches/serial/linked/Project.toml`
  No Changes to `~/GCBenchmarks/benches/serial/linked/Manifest.toml`
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │      50297 │   48807 │     41368 │       7439 │         2172 │                 6 │     4034 │         97 │
│  median │      54143 │   52600 │     44078 │       8607 │         3089 │                 9 │     4048 │         97 │
│ maximum │      89575 │   88079 │     74315 │      13758 │         4959 │                27 │     4072 │         98 │
│   stdev │      11430 │   11436 │      9714 │       1845 │          928 │                 8 │       14 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

x86 server:
```
category = "linked"
bench = "list.jl"
  No Changes to `~/GCBenchmarks/benches/serial/linked/Project.toml`
  No Changes to `~/GCBenchmarks/benches/serial/linked/Manifest.toml`
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       7205 │    4740 │      3419 │       1320 │         1625 │                 6 │     2557 │         63 │
│  median │       7455 │    5015 │      3637 │       1369 │         1807 │                 7 │     2600 │         66 │
│ maximum │       8035 │    5657 │      4266 │       1416 │         2395 │                10 │     2622 │         69 │
│   stdev │        300 │     305 │       275 │         39 │          248 │                 2 │       20 │          2 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

For comparison, on other benchmarks, the M1 is faster:
```
category = "append"
bench = "append.jl"
  No Changes to `~/GCBenchmarks/benches/serial/append/Project.toml`
  No Changes to `~/GCBenchmarks/benches/serial/append/Manifest.toml`
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │        888 │     127 │        52 │         73 │           30 │                 5 │     1483 │          8 │
│  median │        907 │     132 │        56 │         76 │           33 │                 7 │     1484 │          9 │
│ maximum │       1163 │     145 │        65 │         84 │           42 │                12 │     1485 │         10 │
│   stdev │        109 │       6 │         4 │          3 │            4 │                 2 │        0 │          1 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

vs
```
category = "append"
bench = "append.jl"
  No Changes to `~/GCBenchmarks/benches/serial/append/Project.toml`
  No Changes to `~/GCBenchmarks/benches/serial/append/Manifest.toml`
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       2305 │     108 │        84 │         24 │           48 │                 3 │     1483 │          1 │
│  median │       2664 │     112 │        87 │         25 │           49 │                 4 │     1484 │          2 │
│ maximum │       3442 │     139 │       102 │         40 │           59 │                 4 │     1524 │          2 │
│   stdev │        345 │      11 │         7 │          5 │            5 │                 1 │       16 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

Benchmarks are from https://github.com/JuliaCI/GCBenchmarks.

**Comment by vchuravy** [comment]
I assume this is on master? cc: @d-netto

**Comment by Keno** [comment]
Yes, this was master.

**Comment by gbaraldi** [comment]
```julia
bench = "list.jl"
  No Changes to `~/GCBenchmarks/benches/serial/linked/Project.toml`
  No Changes to `~/GCBenchmarks/benches/serial/linked/Manifest.toml`
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       9253 │    6099 │      5380 │        704 │         1981 │                12 │     3055 │         63 │
│  median │       9342 │    6118 │      5396 │        722 │         1995 │                13 │     3136 │         64 │
│ maximum │       9424 │    6226 │      5503 │        725 │         2031 │                18 │     3140 │         65 │
│   stdev │         62 │      53 │        51 │          6 │           19 │                 2 │       26 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
Julia Version 1.9.0-beta3
Commit 24204a73447 (2023-01-18 07:20 UTC)
Platform Info:
  OS: Linux (aarch64-linux-gnu)
  CPU: 4 × Neoverse-N1
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, neoverse-n1)
  Threads: 1 on 4 virtual cores
```
Doesn't seem to be aarch64 related. It would be nice to test on x86 mac as well.

**Comment by d-netto** [comment]
`linked/list.jl` got worse after #47292 on M1 (`master` refers before the merge and `PR` after):

```
┌─────────────────┬────────┬────────┐
│                 │ master │     PR │
├─────────────────┼────────┼────────┤
│ total time [ms] │  76984 │ 102113 │
│    gc time [ms] │  75482 │ 101252 │
│  mark time [ms] │  53561 │  73611 │
│ sweep time [ms] │  21905 │  27691 │
│  max pause [ms] │   4038 │   4330 │
│ max memory [MB] │   4051 │   4047 │
│          pct gc │     98 │     99 │
└─────────────────┴────────┴────────┘
```

The same benchmark had an opposite trend on a AMD/x86-64 machine:

```
┌─────────────────┬────────┬──────┐
│                 │ master │   PR │
├─────────────────┼────────┼──────┤
│ total time [ms] │   7152 │ 5736 │
│    gc time [ms] │   5357 │ 3918 │
│  mark time [ms] │   4826 │ 3376 │
│ sweep time [ms] │    532 │  542 │
│  max pause [ms] │   1787 │ 1302 │
│ max memory [MB] │   2410 │ 2410 │
│          pct gc │     74 │   67 │
└─────────────────┴────────┴──────┘
```

**Comment by Keno** [comment]
Would be interesting to test M1 linux as well. I guess a VM might be fine to just get a rough idea.

**Comment by gbaraldi** [comment]
```julia

bench = "list.jl"
  No Changes to `~/gctest/Resources/julia/bin/GCBenchmarks/benches/serial/linked/Project.toml`
  No Changes to `~/gctest/Resources/julia/bin/GCBenchmarks/benches/serial/linked/Manifest.toml`
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬──────
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ tim ⋯
│         │         ms │      ms │        ms │         ms │           ms │     ⋯
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼──────
│ minimum │       7592 │    5487 │      4825 │        662 │         1792 │     ⋯
│  median │       7912 │    5772 │      5075 │        695 │         1888 │     ⋯
│ maximum │       8352 │    6050 │      5339 │        720 │         1982 │     ⋯
│   stdev │        266 │     184 │       173 │         19 │           69 │     ⋯
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴──────
                                                               3 columns omitted

bench = "append.jl"
  No Changes to `~/gctest/Resources/julia/bin/GCBenchmarks/benches/serial/append/Project.toml`
  No Changes to `~/gctest/Resources/julia/bin/GCBenchmarks/benches/serial/append/Manifest.toml`
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬──────
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ tim ⋯
│         │         ms │      ms │        ms │         ms │           ms │     ⋯
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼──────
│ minimum │       2167 │     240 │        86 │        153 │           58 │     ⋯
│  median │       2222 │     251 │        90 │        162 │           63 │     ⋯
│ maximum │       2356 │     263 │        95 │        172 │           65 │     ⋯
│   stdev │         55 │       7 │         3 │          7 │            2 │     ⋯
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴──────


julia> versioninfo()
Julia Version 1.9.0-beta3
Commit 24204a73447 (2023-01-18 07:20 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin21.4.0)
  CPU: 12 × Intel(R) Core(TM) i7-8700B CPU @ 3.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-14.0.6 (ORCJIT, skylake)
  Threads: 1 on 12 virtual cores
Environment:
  DYLD_FALLBACK_LIBRARY_PATH = /Users/julia/lib:/usr/local/lib:/lib:/usr/lib:/Users/julia/Library/Python/3.7/lib
  JULIA_PKG_PRECOMPILE_AUTO = 1
  JULIA_PKG_SERVER = https://pkg.julialang.org
```
It seems to be m1 related 🤔

**Comment by gbaraldi** [comment]
Mac is doing something really bad here because this is on a VM in my m1 laptop
<img width="969" alt="image" src="https://user-images.githubusercontent.com/28694980/215867085-6003cd7d-d4e2-4341-84ae-9d24b6ea2112.png">
I couldn't get the clipboard working so bear witb the prints

**Comment by Keno** [comment]
It seems surprising that there would be such a big OS difference here .... We're not really doing any system calls. I guess we might be stressing the memory subsystem, but jeez....

**Comment by Keno** [comment]
I did a bare metal linux run and the results match @gbaraldi's VM results.

**Comment by d-netto** [comment]
For reference, the benchmark is inserting nodes into a linked list:

```
mutable struct ListNode
  key::Int64
  next::ListNode
  ListNode() = new()
  ListNode(x)= new(x)
  ListNode(x,y) = new(x,y);
end

function list(n=128)
    start::ListNode = ListNode(1)
    current::ListNode = start
    for i = 2:(n*1024^2)
        current = ListNode(i,current)
    end
    return current.key
end
```

It seems that the dependency with the list size/heap size is quite different on M1/macOS compared to x86-64/linux (more data points needed):

x86-64/linux:

* n = 32

```
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       1835 │    1361 │      1179 │        180 │          428 │                21 │      801 │         70 │
│    mean │       1850 │    1375 │      1191 │        184 │          431 │                23 │      808 │         70 │
│ maximum │       1858 │    1384 │      1198 │        192 │          435 │                25 │      810 │         70 │
│   stdev │          6 │       6 │         6 │          4 │            2 │                 1 │        4 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

* n = 128

```
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       6554 │    4412 │      3852 │        545 │         1445 │                21 │     2704 │         66 │
│    mean │       6603 │    4463 │      3900 │        563 │         1465 │                25 │     2730 │         66 │
│ maximum │       6635 │    4530 │      3956 │        583 │         1509 │                32 │     2733 │         67 │
│   stdev │         24 │      32 │        28 │         11 │           18 │                 4 │        9 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

M1/macOS

* n = 32

```
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       3269 │    2918 │      2470 │        449 │          559 │                 5 │      959 │         88 │
│    mean │       3802 │    3471 │      2924 │        546 │          632 │                 8 │      982 │         89 │
│ maximum │       7163 │    6849 │      5809 │       1040 │          696 │                15 │      996 │         95 │
│   stdev │       1197 │    1203 │      1025 │        179 │           49 │                 3 │       13 │          2 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

* n = 128

```
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │     155148 │  153509 │    108680 │      44815 │         4837 │                 4 │     4034 │         99 │
│    mean │     172498 │  170880 │    120176 │      50666 │         4889 │                 6 │     4038 │         99 │
│ maximum │     203540 │  201985 │    139791 │      62129 │         4936 │                 9 │     4043 │         99 │
│   stdev │      26945 │   26999 │     17071 │       9928 │           49 │                 2 │        4 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

**Comment by Keno** [comment]
GC log on M1/macOS:
```
GC: pause 6.97ms. collected 38.484309MB. incr
GC: pause 29.97ms. collected 0.093800MB. incr
GC: pause 49.96ms. collected 0.000000MB. incr
GC: pause 77.46ms. collected 0.000000MB. incr
GC: pause 109.58ms. collected 0.000000MB. incr
GC: pause 161.82ms. collected 0.000000MB. incr
GC: pause 241.54ms. collected 0.000000MB. incr
GC: pause 369.79ms. collected 0.000000MB. incr
GC: pause 563.72ms. collected 0.000000MB. incr
GC: pause 790.89ms. collected 0.000000MB. incr
GC: pause 729.75ms. collected 0.000000MB. full
GC: pause 1206.77ms. collected 14.342874MB. full
GC: pause 1235.39ms. collected 2.383547MB. full
GC: pause 1272.20ms. collected 0.000000MB. full
GC: pause 1284.24ms. collected 0.000000MB. full
GC: pause 1305.48ms. collected 0.000000MB. full
GC: pause 1328.40ms. collected 0.000000MB. full
GC: pause 1349.18ms. collected 0.000000MB. full
GC: pause 1369.07ms. collected 0.000000MB. full
GC: pause 1395.83ms. collected 0.000000MB. full
GC: pause 1420.44ms. collected 0.000000MB. full
GC: pause 1438.30ms. collected 0.000000MB. full
GC: pause 1462.76ms. collected 0.000000MB. full
GC: pause 1487.71ms. collected 0.000000MB. full
GC: pause 1512.28ms. collected 0.000000MB. full
GC: pause 1531.34ms. collected 0.000000MB. full
GC: pause 1553.63ms. collected 0.000000MB. full
GC: pause 1571.83ms. collected 0.000000MB. full
GC: pause 1597.79ms. collected 0.000000MB. full
GC: pause 1617.82ms. collected 0.000000MB. full
GC: pause 1634.78ms. collected 0.000000MB. full
GC: pause 1657.50ms. collected 0.000000MB. full
GC: pause 1683.11ms. collected 0.000000MB. full
GC: pause 1704.77ms. collected 0.000000MB. full
GC: pause 1729.11ms. collected 0.000000MB. full
GC: pause 1934.86ms. collected 0.000000MB. full
GC: pause 1768.93ms. collected 0.000000MB. full
GC: pause 1794.73ms. collected 0.000000MB. full
GC: pause 1830.96ms. collected 0.000000MB. full
GC: pause 1844.74ms. collected 0.000000MB. full
GC: pause 1870.58ms. collected 0.000000MB. full
GC: pause 1893.89ms. collected 0.000000MB. full
GC: pause 1900.76ms. collected 0.000000MB. full
GC: pause 1913.94ms. collected 0.000000MB. full
GC: pause 1937.05ms. collected 0.000000MB. full
GC: pause 1961.40ms. collected 0.000000MB. full
GC: pause 1979.99ms. collected 0.000000MB. full
GC: pause 2001.17ms. collected 0.000000MB. full
GC: pause 2036.07ms. collected 0.000000MB. full
GC: pause 2048.52ms. collected 0.000000MB. full
GC: pause 2104.55ms. collected 0.000000MB. full
GC: pause 2120.38ms. collected 0.000000MB. full
GC: pause 2135.64ms. collected 13.838608MB. full
(value = 134217728, times = 0x00000010f92eb423, gc_diff = Base.GC_Diff(4294967296, 0, 0, 134217728, 0, 175, 71423721960, 52, 42), gc_end = Base.GC_Num(30364928, 0, 0, 1483, 1, 142908647, 50668, 1480, 71518025462, 5312867266, 0, 0x0000000002bc0000, 74, 42, 2120381292, 4244861763, 7542, 15209, 322084250, 1798285791, 10952903502, 60563773457))
```

vs x86_64 linux:
```
GC: pause 18.85ms. collected 26.226526MB. incr
GC: pause 41.31ms. collected 0.146512MB. incr
GC: pause 65.38ms. collected 0.000000MB. incr
GC: pause 83.65ms. collected 0.000000MB. incr
GC: pause 155.24ms. collected 0.000000MB. incr
GC: pause 174.23ms. collected 0.000000MB. incr
GC: pause 260.61ms. collected 0.000000MB. incr
GC: pause 394.58ms. collected 0.000000MB. incr
GC: pause 617.96ms. collected 0.000000MB. incr
GC: pause 1018.14ms. collected 0.000000MB. incr
GC: pause 1329.41ms. collected 0.000000MB. incr
(value = 134217728, times = 0x0000000158a0571d, gc_diff = Base.GC_Diff(4294967296, 0, 0, 134217728, 0, 13, 4159353906, 11, 0), gc_end = Base.GC_Num(822962816, 0, 0, 35, 0, 136652179, 1193, 27, 4230891708, 3642918580, 0, 0x0000000068707a6b, 14, 0, 1329409486, 2725647287, 15250, 20040, 280061775, 1049347211, 872873882, 3357979326))
```

For some reason on macOS, we're just doing way more collections and in particular an excessive amount of full collections.

**Comment by KristofferC** [comment]
What's the status on 1.8?

**Comment by Keno** [comment]
Looks like 1.8 is fine:
```
GC: pause 11.79ms. collected 24.086122MB. incr
GC: pause 26.25ms. collected 0.008432MB. incr
GC: pause 47.50ms. collected 0.003504MB. full
GC: pause 124.23ms. collected 0.875768MB. full
GC: pause 219.33ms. collected 0.000000MB. full
GC: pause 488.75ms. collected 0.000000MB. full
GC: pause 1290.01ms. collected 0.000000MB. full
(value = 134217728, times = 0x00000000b6d8db62, gc_diff = Base.GC_Diff(4294967296, 0, 0, 134217728, 0, 20, 2207871376, 7, 5), gc_end = Base.GC_Num(1270836832, 0, 0, 210, 0, 136860540, 1828, 208, 2230395626, 3231358286, 0, 0x000000010b076000, 11, 5))
```

**Comment by d-netto** [comment]
Could to be related to https://github.com/JuliaLang/julia/pull/44805.

For reference, after commenting out:

```c
// If the live data outgrows the suggested max_total_memory
// we keep going with minimum intervals and full gcs until
// we either free some space or get an OOM error.
if (live_bytes > max_total_memory) {
    sweep_full = 1;
}
```

and:

```c
// We need this for 32 bit but will be useful to set limits on 64 bit
if (gc_num.interval + live_bytes > max_total_memory) {
    if (live_bytes < max_total_memory) {
        gc_num.interval = max_total_memory - live_bytes;
    } else {
        // We can't stay under our goal so let's go back to
        // the minimum interval and hope things get better
        gc_num.interval = default_collect_interval;
   }
}
```

on master it goes to

```
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       6085 │    4596 │      3853 │        728 │         1859 │                 8 │     2740 │         75 │
│    mean │       6235 │    4733 │      3943 │        789 │         2133 │                12 │     2741 │         75 │
│ maximum │       6367 │    4862 │      3986 │        876 │         2312 │                17 │     2745 │         76 │
│   stdev │        101 │      96 │        54 │         60 │          180 │                 4 │        2 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

on the M1/macOS.

**Comment by gbaraldi** [comment]
I wonder if we're getting the wrong values from libuv here.

**Comment by oscardssmith** [comment]
are we detecting max memory incorrectly on mac?

**Comment by gbaraldi** [comment]
Yes, `Sys.free_memory()` shows I have 350 MB free while I have over 5GB

**Comment by d-netto** [comment]
Seems like it's the case: hardcoding 2GB (which is closer to what I have on my machine) into `free_mem` (instead of `uv_get_available_memory`) gives me:

```
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       6983 │    5537 │      4367 │       1091 │         2457 │                 8 │     2740 │         78 │
│    mean │       7100 │    5615 │      4466 │       1149 │         2511 │                11 │     2741 │         78 │
│ maximum │       7247 │    5700 │      4546 │       1226 │         2579 │                15 │     2745 │         78 │
│   stdev │        114 │      75 │        70 │         57 │           51 │                 3 │        2 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```

on M1/macOS.

**Comment by gbaraldi** [comment]
Free memory on macos or at least on the M1 doesn't seem to be a reliable way of checking.

**Comment by gbaraldi** [comment]
With https://github.com/JuliaLang/libuv/pull/34 I get
```julia
bench = "list.jl"
┌─────────┬────────────┬─────────┬───────────┬────────────┬──────────────┬───────────────────┬──────────┬────────────┐
│         │ total time │ gc time │ mark time │ sweep time │ max GC pause │ time to safepoint │ max heap │ percent gc │
│         │         ms │      ms │        ms │         ms │           ms │                us │       MB │          % │
├─────────┼────────────┼─────────┼───────────┼────────────┼──────────────┼───────────────────┼──────────┼────────────┤
│ minimum │       5112 │    3703 │      3223 │        466 │         1222 │                 7 │     2701 │         71 │
│  median │       5150 │    3747 │      3262 │        488 │         1236 │                 9 │     2701 │         72 │
│ maximum │       5270 │    3868 │      3372 │        507 │         1323 │                15 │     2705 │         72 │
│   stdev │         57 │      50 │        42 │         12 │           34 │                 2 │        2 │          0 │
└─────────┴────────────┴─────────┴───────────┴────────────┴──────────────┴───────────────────┴──────────┴────────────┘
```
Which seems more reasonable.

**Comment by d-netto** [comment]
Is this on master or 1.9? (not sure if the there would be that much difference between them)

**Comment by vtjnash** [comment]
Duplicate of https://github.com/JuliaLang/julia/issues/47684

**Comment by JeffBezanson** [comment]
Should we be calling `uv_get_available_memory` at all? It might make sense just to use `uv_get_constrained_memory` as a hint of roughly how much memory is ok to use and let swap handle it.

**Comment by vchuravy** [comment]
> Should we be calling `uv_get_available_memory` at all? It might make sense just to use `uv_get_constrained_memory` as a hint of roughly how much memory is ok to use and let swap handle it.

I am in favor of that. It's odd to take the state of the system when Julia starts as the high-water mark. Constrained memory seems like the right concept (in particular since cgroups is a thing).

**Comment by gbaraldi** [comment]
Macos has no concept of constrained memory, and returns 0 here.

**Comment by gbaraldi** [comment]
After some discussion should we just remove https://github.com/JuliaLang/julia/blob/d72a9a1d2a2bd181cabd5294f21ef3242dd09efd/src/gc.c#L3258-L3262 and change to just using 50% or some other number of the constrained memory? Which probably is total memory on non linux systems?
